clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -265,9 +259,68 @@ public FSDataOutputStream append(Path path, int i, Progressable progressable) th
     }
 
     @Override
-    public boolean rename(Path path, Path path1) throws IOException {
-        LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ rename $$$$$$$$$$$$$$$$$$$$$$$$$$$$ ");
-        return false;
+    public boolean rename(Path src, Path dst) throws IOException {
+        LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ Rename path {} to {} $$$$$$$$$$$$$$$$$$$$$$$$$$$$", src, dst);
+
+        // This implementation supports rename when src and dst are files. directories are not supported. TODO (Tals): support dirs
+
+        FileStatus srcStatus;
+        try {
+            srcStatus = getFileStatus(src);
+        } catch (FileNotFoundException e) {
+            LOG.error("rename: src not found {}", src);
+            return false;
+        }
+
+        ObjectLocation srcObjectLoc = pathToObjectLocation(src);
+        ObjectLocation dstObjectLoc = pathToObjectLocation(dst);
+        if (srcObjectLoc.equals(dstObjectLoc)) {
+            LOG.debug("rename: src and dst refer to the same lakefs object location: {}", dst);
+            return srcStatus.isFile();
+        }
+
+        ObjectStats stats;
+        try {
+            // Stat src file to get its metadata
+            stats = getObjectStats(src);
```

Comment:
I think we just stat'ed this file.  Doing it twice is usually worse than doing it once.  Maybe it existed here and got deleted by the time we opened?  Unless we can offer more information using this, I see no advantage in having *two* paths to return the same error, especially given that we now add a second API call.

Reply:
I removed the first call to getFileStatus following the comment in https://github.com/treeverse/lakeFS/pull/1972/files#r635169771